### PR TITLE
folder_branch_ops: limit background batch sizes, and make batch values more configurable

### PR DESCRIPTION
**Pull request by strib**
This changes the background flusher to flush either a) after a certain amount of time has passed since the last write, or b) after a certain number of directory operations have been buffered.  This should have no effect since we don't currently buffer directory operations (that's what the next PR in the series will do.)

It also makes both these values configurable from the command line, to make performance testing easier.

Issue: KBFS-2075

### CODE REVIEW

File: libkbfs/folder_branch_ops.go [hunk 1]
```diff
@@ -5447,31 +5448,54 @@ func (fbo *folderBranchOps) backgroundFlusher(betweenFlushes time.Duration) {
...
 			doSelect = false
+		} else if fbo.getCachedDirOpsCount(lState) >=
+			fbo.config.BGFlushDirOpBatchSize() {
+			doSelect = false
 		}
 
 		if doSelect {
 			// Wait until we really have a write waiting.
 			forced := false
 			select {
```
Comment (songgao) [inline libkbfs/folder_branch_ops.go]: It seems this `select` block isn't needed anymore?

File: libkbfs/folder_branch_ops.go [hunk 2]
```diff
@@ -5447,31 +5448,54 @@ func (fbo *folderBranchOps) backgroundFlusher(betweenFlushes time.Duration) {
...
 				return
 			}
 
 			if !forced {
-				select {
-				case <-ticker.C:
-				case <-fbo.forceSyncChan:
-				case <-fbo.shutdownChan:
-					return
+				timer := time.NewTimer(fbo.config.BGFlushPeriod())
```
Comment (songgao) [inline libkbfs/folder_branch_ops.go]: `time.Timer` discards unreceived ticks. What do you think about keeping a `timer` outside the out-most for loop so that we don't have to keep constructing new ones?

Comment (songgao) [inline libkbfs/folder_branch_ops.go]: Although I guess that means we'd always get a tick after a slow flush.

File: libkbfs/init.go [hunk 3]
```diff
@@ -158,7 +167,9 @@ func DefaultInitParams(ctx Context) InitParams {
 		},
 		TLFJournalBackgroundWorkStatus: TLFJournalBackgroundWorkEnabled,
 		StorageRoot:                    ctx.GetDataDir(),
-		Mode:                           InitDefaultString,
+		BGFlushPeriod:                  1 * time.Second,
```
Comment (songgao) [inline libkbfs/init.go]: Did you mean to use `bgFlushPeriodDefault` here? (also the `BGFlushDirOpBatchSize` below)

File: libkbfs/kbfs_ops_test.go [hunk 4]
```diff
@@ -4479,7 +4479,8 @@ func TestKBFSOpsBackgroundFlush(t *testing.T) {
 
 	// start the background flusher
 	ops := getOps(config, rootNode.GetFolderBranch().Tlf)
-	go ops.backgroundFlusher(1 * time.Millisecond)
+	config.SetBGFlushPeriod(1 * time.Millisecond)
```
Comment (jzila) [inline libkbfs/kbfs_ops_test.go]: Probably overkill, but maybe we should be using a fake clock here. I won't push on whatever you decide though.

File: libkbfs/folder_branch_ops.go (hunk 1 above)
Comment (strib) [inline libkbfs/folder_branch_ops.go]: We still need a select loop without `<-timer.C` to avoid waking up every 1 second when there are no writes.  You still want to avoid that, right?

File: libkbfs/folder_branch_ops.go (hunk 2 above)
Comment (strib) [inline libkbfs/folder_branch_ops.go]: I did it this way on purpose to make sure the timer started exactly after the first write was signaled.  Also, a timer is use-once, so I think you are suggesting going back to a ticker.  I'd rather keep the timer so that it's measured exactly from the first write signal.  There's not much overhead in this way, since it's only constructed whenever write activity is actually happening.

File: libkbfs/kbfs_ops_test.go (hunk 4 above)
Comment (strib) [inline libkbfs/kbfs_ops_test.go]: I agree -- this test was written before I knew about clockwork, and I didn't have an easy way to use our fake clock with the `time` package.  But since this is an unrelated pre-existing condition, I'm going to leave that change out of this PR.

File: libkbfs/folder_branch_ops.go (hunk 1 above)
Comment (songgao) [inline libkbfs/folder_branch_ops.go]: Ah I see; that makes sense! Then I guess `forced` can be renamed into something like `doNotWait`

File: libkbfs/folder_branch_ops.go (hunk 2 above)
Comment (songgao) [inline libkbfs/folder_branch_ops.go]: OK; sounds good!

File: libkbfs/init.go (hunk 3 above)
Comment (strib) [inline libkbfs/init.go]: Done, thanks.

File: libkbfs/folder_branch_ops.go (hunk 1 above)
Comment (strib) [inline libkbfs/folder_branch_ops.go]: Sounds good -- I renamed it `doWait` and inverted the logic, to avoid having a double negative condition.

File: libkbfs/folder_branch_ops.go [hunk 5]
```diff
@@ -5449,31 +5450,54 @@ func (fbo *folderBranchOps) backgroundFlusher(betweenFlushes time.Duration) {
...
+							fbo.config.BGFlushDirOpBatchSize() {
+							break loop
+						}
+					case <-fbo.forceSyncChan:
+						break loop
+					case <-fbo.shutdownChan:
+						return
+					}
 				}
+				timer.Stop()
```
Comment (akalin-keybase) [inline libkbfs/folder_branch_ops.go]: Don't need this, I think -- it's only tickers that have resources that need to be freed.
